[CN] Block should allow omitting C-type like owned

Owned can omit C types in this context and Block is just an unintialised Owned, so it should allow it too.

```sh
13:16 ➜  cerberus git:(master) cat block_no_annot.c && cn verify block_no_annot.c
void f(int *p)
/*@ requires take V = Block(p);
    ensures  take V2 = Block(p);
@*/
{ }
block_no_annot.c:2:28: error: unexpected token after 'Block' and before '('
parsing "pred": seen "CN_BLOCK", expecting "LT ctype GT"
/*@ requires take V = Block(p);
                           ^
```

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -1594,7 +1594,10 @@ let rec check_expr labels (e : BT.t mu_expr) (k : IT.t -> unit m) : unit m =
              | E_Pred (CN_owned (Some ct)) ->
                let@ () = WellTyped.WCT.is_ct loc ct in
                return (ResourceTypes.Owned (ct, Init))
-             | E_Pred (CN_block ct) ->
+             | E_Pred (CN_block None) ->
+               let msg = "'extract' requires a C-type annotation for 'Block'" in
+               fail (fun _ -> { loc; msg = Generic !^msg })
+             | E_Pred (CN_block (Some ct)) ->
                let@ () = WellTyped.WCT.is_ct loc ct in
                return (ResourceTypes.Owned (ct, Uninit))
              | E_Pred (CN_named pn) ->

--- a/backend/cn/lib/cnprog.ml
+++ b/backend/cn/lib/cnprog.ml
@@ -93,7 +93,7 @@ let dtree_of_to_extract = function
     let pred =
       match pred with
       | CN_owned oct -> CN_owned (Option.map Sctypes.to_ctype oct)
-      | CN_block ct -> CN_block (Sctypes.to_ctype ct)
+      | CN_block ct -> CN_block (Option.map Sctypes.to_ctype ct)
       | CN_named p -> CN_named p
     in
     Dnode (pp_ctor "[CN]pred", [ Cerb_frontend.Cn_ocaml.PpAil.dtree_of_cn_pred pred ])

--- a/backend/cn/lib/compile.ml
+++ b/backend/cn/lib/compile.ml
@@ -1031,7 +1031,7 @@ module EffectfulTranslation = struct
       | x :: xs -> return (x, xs)
     in
     let@ pname, oargs_ty =
-      let infer_scty cname oty = 
+      let infer_scty cname oty =
         match oty with
         | Some ty -> return (Sctypes.of_ctype_unsafe res_loc ty)
         | None ->
@@ -1042,15 +1042,11 @@ module EffectfulTranslation = struct
                { loc;
                  msg =
                    Generic
-                     ( !^"Cannot tell C-type of pointer. Please use " 
-                       ^^^
-                       !^cname 
-                       ^^^ 
-                       !^" with an annotation: \'"
-                       ^^^ 
-                       !^cname 
-                       ^^^
-                       !^"<CTYPE>'." )
+                     (!^"Cannot tell C-type of pointer. Please use "
+                      ^^^ !^cname
+                      ^^^ !^" with an annotation: \'"
+                      ^^^ !^cname
+                      ^^^ !^"<CTYPE>'.")
                }
            | has ->
              let expected = "pointer" in
@@ -1060,7 +1056,8 @@ module EffectfulTranslation = struct
                  msg =
                    Illtyped_it
                      { it = Terms.pp ptr_expr; has = SBT.pp has; expected; reason }
-               }) in 
+               })
+      in
       match res with
       | CN_owned oty ->
         let@ scty = infer_scty "Owned" oty in
@@ -1638,8 +1635,8 @@ let translate_cn_statement
          | E_Everything -> E_Everything
          | E_Pred (CN_owned oty) ->
            E_Pred (CN_owned (Option.map (Sctypes.of_ctype_unsafe loc) oty))
-         | E_Pred (CN_block oty) -> 
-            E_Pred (CN_block (Option.map (Sctypes.of_ctype_unsafe loc) oty))
+         | E_Pred (CN_block oty) ->
+           E_Pred (CN_block (Option.map (Sctypes.of_ctype_unsafe loc) oty))
          | E_Pred (CN_named pn) -> E_Pred (CN_named pn)
        in
        return (M_CN_statement (loc, M_CN_extract (attrs, to_extract, expr)))

--- a/backend/cn/lib/compile.ml
+++ b/backend/cn/lib/compile.ml
@@ -1061,9 +1061,15 @@ module EffectfulTranslation = struct
            information *)
         let oargs_ty = Memory.sbt_of_sct scty in
         return (Owned (scty, Init), oargs_ty)
-      | CN_block ty ->
-        let scty = Sctypes.of_ctype_unsafe res_loc ty in
-        return (Owned (scty, Uninit), Memory.sbt_of_sct scty)
+      | CN_block oty ->
+          (match oty with 
+          | None -> 
+            fail { loc; msg = Generic
+                       !^"Cannot tell C-type of pointer. Please use Block with an \
+                          annotation: \'Block<CTYPE>'." }
+          | Some ty -> 
+            let scty = Sctypes.of_ctype_unsafe res_loc ty in
+            return (Owned (scty, Uninit), Memory.sbt_of_sct scty))
       | CN_named pred ->
         let@ pred_sig =
           match lookup_predicate pred env with
@@ -1630,7 +1636,8 @@ let translate_cn_statement
          | E_Everything -> E_Everything
          | E_Pred (CN_owned oty) ->
            E_Pred (CN_owned (Option.map (Sctypes.of_ctype_unsafe loc) oty))
-         | E_Pred (CN_block ty) -> E_Pred (CN_block (Sctypes.of_ctype_unsafe loc ty))
+         | E_Pred (CN_block oty) -> 
+            E_Pred (CN_block (Option.map (Sctypes.of_ctype_unsafe loc) oty))
          | E_Pred (CN_named pn) -> E_Pred (CN_named pn)
        in
        return (M_CN_statement (loc, M_CN_extract (attrs, to_extract, expr)))

--- a/backend/cn/lib/compile.ml
+++ b/backend/cn/lib/compile.ml
@@ -1062,14 +1062,35 @@ module EffectfulTranslation = struct
         let oargs_ty = Memory.sbt_of_sct scty in
         return (Owned (scty, Init), oargs_ty)
       | CN_block oty ->
-          (match oty with 
-          | None -> 
-            fail { loc; msg = Generic
+        (* TODO: deduplicate wrt CN_owned above *)
+        let@ scty =
+          match oty with
+          | Some ty -> return (Sctypes.of_ctype_unsafe res_loc ty)
+          | None ->
+            (match IT.bt ptr_expr with
+             | SBT.Loc (Some ty) -> return ty
+             | Loc None ->
+               fail
+                 { loc;
+                   msg =
+                     Generic
                        !^"Cannot tell C-type of pointer. Please use Block with an \
-                          annotation: \'Block<CTYPE>'." }
-          | Some ty -> 
-            let scty = Sctypes.of_ctype_unsafe res_loc ty in
-            return (Owned (scty, Uninit), Memory.sbt_of_sct scty))
+                          annotation: \'Block<CTYPE>'."
+                 }
+             | has ->
+               let expected = "pointer" in
+               let reason = "Block<_> predicate" in
+               fail
+                 { loc;
+                   msg =
+                     Illtyped_it
+                       { it = Terms.pp ptr_expr; has = SBT.pp has; expected; reason }
+                 })
+        in
+        (* we don't take Resources.owned_oargs here because we want to maintain the C-type
+           information *)
+        let oargs_ty = Memory.sbt_of_sct scty in
+        return (Owned (scty, Uninit), oargs_ty)
       | CN_named pred ->
         let@ pred_sig =
           match lookup_predicate pred env with

--- a/backend/cn/lib/compile.ml
+++ b/backend/cn/lib/compile.ml
@@ -1031,64 +1031,45 @@ module EffectfulTranslation = struct
       | x :: xs -> return (x, xs)
     in
     let@ pname, oargs_ty =
+      let infer_scty cname oty = 
+        match oty with
+        | Some ty -> return (Sctypes.of_ctype_unsafe res_loc ty)
+        | None ->
+          (match IT.bt ptr_expr with
+           | SBT.Loc (Some ty) -> return ty
+           | Loc None ->
+             fail
+               { loc;
+                 msg =
+                   Generic
+                     ( !^"Cannot tell C-type of pointer. Please use " 
+                       ^^^
+                       !^cname 
+                       ^^^ 
+                       !^" with an annotation: \'"
+                       ^^^ 
+                       !^cname 
+                       ^^^
+                       !^"<CTYPE>'." )
+               }
+           | has ->
+             let expected = "pointer" in
+             let reason = cname ^ "<_> predicate" in
+             fail
+               { loc;
+                 msg =
+                   Illtyped_it
+                     { it = Terms.pp ptr_expr; has = SBT.pp has; expected; reason }
+               }) in 
       match res with
       | CN_owned oty ->
-        let@ scty =
-          match oty with
-          | Some ty -> return (Sctypes.of_ctype_unsafe res_loc ty)
-          | None ->
-            (match IT.bt ptr_expr with
-             | SBT.Loc (Some ty) -> return ty
-             | Loc None ->
-               fail
-                 { loc;
-                   msg =
-                     Generic
-                       !^"Cannot tell C-type of pointer. Please use Owned with an \
-                          annotation: \'Owned<CTYPE>'."
-                 }
-             | has ->
-               let expected = "pointer" in
-               let reason = "Owned<_> predicate" in
-               fail
-                 { loc;
-                   msg =
-                     Illtyped_it
-                       { it = Terms.pp ptr_expr; has = SBT.pp has; expected; reason }
-                 })
-        in
+        let@ scty = infer_scty "Owned" oty in
         (* we don't take Resources.owned_oargs here because we want to maintain the C-type
            information *)
         let oargs_ty = Memory.sbt_of_sct scty in
         return (Owned (scty, Init), oargs_ty)
       | CN_block oty ->
-        (* TODO: deduplicate wrt CN_owned above *)
-        let@ scty =
-          match oty with
-          | Some ty -> return (Sctypes.of_ctype_unsafe res_loc ty)
-          | None ->
-            (match IT.bt ptr_expr with
-             | SBT.Loc (Some ty) -> return ty
-             | Loc None ->
-               fail
-                 { loc;
-                   msg =
-                     Generic
-                       !^"Cannot tell C-type of pointer. Please use Block with an \
-                          annotation: \'Block<CTYPE>'."
-                 }
-             | has ->
-               let expected = "pointer" in
-               let reason = "Block<_> predicate" in
-               fail
-                 { loc;
-                   msg =
-                     Illtyped_it
-                       { it = Terms.pp ptr_expr; has = SBT.pp has; expected; reason }
-                 })
-        in
-        (* we don't take Resources.owned_oargs here because we want to maintain the C-type
-           information *)
+        let@ scty = infer_scty "Block" oty in
         let oargs_ty = Memory.sbt_of_sct scty in
         return (Owned (scty, Uninit), oargs_ty)
       | CN_named pred ->

--- a/backend/cn/lib/wellTyped.ml
+++ b/backend/cn/lib/wellTyped.ml
@@ -1782,7 +1782,8 @@ module BaseTyping = struct
         match to_extract with
         | E_Pred (CN_owned None) -> return ()
         | E_Pred (CN_owned (Some ct)) -> WCT.is_ct loc ct
-        | E_Pred (CN_block ct) -> WCT.is_ct loc ct
+        | E_Pred (CN_block None) -> return ()
+        | E_Pred (CN_block (Some ct)) -> WCT.is_ct loc ct
         | E_Pred (CN_named p) ->
           let@ _ = get_resource_predicate_def loc p in
           return ()

--- a/frontend/model/cabs_to_ail.lem
+++ b/frontend/model/cabs_to_ail.lem
@@ -4322,9 +4322,11 @@ let desugar_cn_pred = function
       E.return (CN_owned (Just ty))
   | CN_owned Nothing ->
       E.return (CN_owned Nothing)
-  | CN_block tyname ->
+  | CN_block (Just tyname) ->
       desugar_type_name tyname >>= fun (_, ty) ->
-      E.return (CN_block ty)
+      E.return (CN_block (Just ty))
+  | CN_block Nothing ->
+      E.return (CN_block Nothing)
   | CN_named ident ->
       E.resolve_cn_ident CN_predicate ident >>= fun sym_resolved ->
       let sym = Cn.ensure_not_c_variable sym_resolved in

--- a/frontend/model/cn.lem
+++ b/frontend/model/cn.lem
@@ -131,7 +131,7 @@ let mk_bool b = CNExpr Loc.unknown (CNExpr_const (CNConst_bool b))
 
 type cn_pred 'a 'ty =
   | CN_owned of maybe 'ty
-  | CN_block of 'ty
+  | CN_block of maybe 'ty
   | CN_named of 'a
 
 type cn_resource 'a 'ty =

--- a/ocaml_frontend/cn_ocaml.ml
+++ b/ocaml_frontend/cn_ocaml.ml
@@ -267,8 +267,10 @@ module MakePp (Conf: PP_CN) = struct
       Dleaf (pp_stmt_ctor "CN_owned" ^^^ Conf.pp_ty ty)
     | CN_owned None ->
       Dleaf (pp_stmt_ctor "CN_owned" ^^^ P.parens !^"no C-type")
-    | CN_block ty ->
+    | CN_block (Some ty) ->
       Dleaf (pp_stmt_ctor "CN_block" ^^^ Conf.pp_ty ty)
+    | CN_block None ->
+      Dleaf (pp_stmt_ctor "CN_block" ^^^ P.parens !^"no C-type")
     | CN_named ident ->
         Dleaf (pp_stmt_ctor "CN_named" ^^^ P.squotes (Conf.pp_ident ident))
 

--- a/parsers/c/c_parser.mly
+++ b/parsers/c/c_parser.mly
@@ -2392,7 +2392,9 @@ pred:
 | CN_OWNED
     { Cerb_frontend.Cn.CN_owned None }
 | CN_BLOCK ty= delimited(LT, ctype, GT)
-    { Cerb_frontend.Cn.CN_block ty }
+    { Cerb_frontend.Cn.CN_block (Some ty) }
+| CN_BLOCK
+    { Cerb_frontend.Cn.CN_block None }
 | str= UNAME VARIABLE
     { Cerb_frontend.Cn.CN_named (Symbol.Identifier (Cerb_location.point $startpos(str), str)) }
 ;

--- a/tests/cn/block_type.c
+++ b/tests/cn/block_type.c
@@ -1,0 +1,7 @@
+void f(int *p)
+/*@ requires take V = Block(p);
+    ensures  take V2 = Block(p);
+@*/
+{ 
+  ; 
+}

--- a/tests/cn/block_type.c
+++ b/tests/cn/block_type.c
@@ -1,7 +1,17 @@
-void f(int *p)
+// Block does not need to take a CTYPE parameter if it can infer the type from the environment
+
+void block_notype_1(int *p)
 /*@ requires take V = Block(p);
     ensures  take V2 = Block(p);
 @*/
 { 
   ; 
+}
+
+void block_notype_2(int *p)
+/*@ requires take V = Block(p);
+    ensures  take V2 = Owned(p);
+@*/
+{ 
+  *p = 7; 
 }

